fix(buffers): tighten up logic around not overrunning maximum buffer size

tobz opened the pull request:
# Context

As mentioned in #13263, the error message around specifying the minimum allowed value for `buffer.max_size`, when using a disk buffer, can be confusing to users upgrading, as there was no lower bound previously. While #13352 improves the situation somewhat by documenting this fact, it's still unnecessarily confusing that users must plan for 256MB of available disk space when they can set `buffer.max_size` as low as 128MB.

# Solution

We've reworked both the configuration builder logic as well as, more fundamentally, how strict/precise we are around obeying the configured limits on the writer side. We primarily do three things differently now with this PR:

- use the actual lower bound (256MB) in the config validation error message
- use a precise calculation for `max_buffer_size` as it's used internally
- disallow writing records bigger than the maximum data file size

## Fixing the error message and the calculated `max_buffer_size` value

While a user may specify `buffer.max_size` as something like, say, 500MB, we use a specific internal accounting of the buffer size in terms of tracking where it is in relation to the maximum allowed size. To wit, we track the current buffer size as the sum of the size of all buffered events which have not yet been read/acknowledged, which means that the sum of the size of all data files on disk could exceed the configured maximum buffer size.

Prior to this PR, we tried to compensate for all of this by doing some math when calculating the maximum buffer size value we'd work off of, which looks like:

```rust
let max_size_actual = max_size_config - (max_size_config % max_data_file_size);
let max_size_actual = max(max_size_actual, max_data_file_size);
```

This ensured that our maximum buffer size was always at least as much as the maximum size of a data file, because since we track only unread/unacknowledged events in the buffer against its current size, we could have a data file that was filled to its maximum of 128MB, with, say, 10MB of events outstanding, and we might _technically_ be under a maximum buffer size limit of 50MB, but we're violating the spirit of that limit in terms of the size of all the buffer data files on disk.

Further, we were attempting to adjust the maximum buffer size value to account for the desire to always be able to start writing to a second data file so that the writer could make progress as the reader made progress with reading/acknowledging events. This is also why the current buffer size is based on remaining events rather than the total size on disk. However, the above calculation (first line) simply sets the maximum buffer size to an increment of the maximum data file size: it doesn't actually do what we want, which is to figure out how much data we can store, including the overhead of an outstanding data file (fully processed, but not yet deleted), without going over the user-specified maximum buffer size.

Instead, we now correctly warn the user that the maximum buffer size must be at least 256MB (2x of the maximum data file size, which is currently hard-coded) and, more importantly, we simply subtract the maximum data file size from the maximum buffer size, which now correctly ensures we leave ourselves enough slack to stay within the maximum buffer size, knowing we may have an outstanding data file, but doesn't lose any usable space by forcing ourselves to align to a multiple of the maximum data file size.

## Fixing our imprecise limits on data file sizes/record sizes

Additionally, but more importantly, we previously had logic that would allow a record to technically create a data file that exceeded the maximum data file size as we did not correct account for the fact that while the event passed into the buffer may encode itself down to N bytes, we actually wrap that event with additional information needed by the buffer, which ends up taking something more like N+48 bytes when all said and done. This represented a potential to violate the contract that the configured maximum buffer size was the actual maximum number of bytes we would attempt to use.

We now correctly compensate for what we call the "record header" -- the metadata that wraps an encoded event -- and all of the minutiae surrounding record serialization, such as overaligned buffers, etc. This is reflected both in the buffer writer logic itself, as well as the configuration validation logic itself.

bruceg commented on lib/vector-buffers/src/variants/disk_v2/tests/invariants.rs:
```diff
@@ -249,10 +255,7 @@ async fn writer_stops_when_hitting_file_that_reader_is_still_on() {
-                let record = SizedRecord(record_size);
-                writer.write_record(record).await
-            });
+            let mut blocked_write = spawn(async { writer.write_record(record).await });
```
Can this be done without the `async` wrapper? It looks like `write_record` is just a future…
```suggestion
            let mut blocked_write = spawn(writer.write_record(record));
```

tobz replied:
Yep, I agree with your assessment.

bruceg commented on lib/vector-buffers/src/variants/disk_v2/tests/invariants.rs:
```diff
@@ -610,13 +614,18 @@ async fn writer_waits_for_reader_after_validate_last_write_fails_and_data_file_s
+            let record_size = 42;
+            let record = SizedRecord(record_size);
+            let corrected_record_size = get_corrected_max_record_size(&record);
+            let max_data_file_size = (corrected_record_size * 2)
+                .try_into()
+                .expect("Vector does not supports 128-bit architectures.");
```
This exception is a bit misleading, as it requires both an 128-bit architecture and a value of at least 2^64, which is rather impossible for a record size of 42.

tobz replied:
Fair. I'll make it more relevant to the context/callsite.

tobz replied:
Addressed in c54ae438f.

bruceg commented on lib/vector-buffers/src/variants/disk_v2/tests/mod.rs:
```diff
@@ -273,8 +305,12 @@ where
     R: Bufferable,
 {
+    let max_record_size =
+        usize::try_from(max_data_file_size).expect("Vector only supports 64-bit architectures.");
```
Is this worth a wrapper function?

tobz replied:
I think in the tests, I'm just gonna keep the casts and switch to `unwrap`... since all a wrapper would end up looking like, I think, is something that resembles `max_data_file_size_to_max_record_size` which is just the same code -- `usize::try_from` -- and feels extraneous to me?

tobz replied:
Addressed in c54ae438f.

bruceg commented on lib/vector-buffers/src/variants/disk_v2/tests/mod.rs:
```diff
@@ -307,3 +344,46 @@ where
         .expect("should not fail to create buffer")
 }
+
+const fn align16(amount: usize) -> usize {
```
This duplicates `align16` above.

tobz replied:
Yep, good point. Will remove.

tobz replied:
Addressed in c54ae438f.

bruceg commented on lib/vector-buffers/src/variants/disk_v2/tests/mod.rs:
```diff
@@ -307,3 +344,46 @@ where
+    }
+}
+
+const fn align16_u32(amount: u32) -> u32 {
```
This cries out for a generic version.

bruceg reviewed (commented):
This makes sense to me. I left some minor suggestions below.

tobz commented on Cargo.lock:
```diff
@@ -6886,9 +6886,9 @@ checksum = "eb703cfe953bccee95685111adeedb76fabe4e97549a58d16f03ea7b9367bb32"
 [[package]]
 name = "smallvec"
-version = "1.8.1"
+version = "1.8.0"
```
Hmmm, not intentional. I'll clean that up.

tobz commented on lib/vector-buffers/src/variants/disk_v2/ledger.rs:
```diff
@@ -705,21 +708,19 @@ where
         f.debug_struct("Ledger")
             .field("config", &self.config)
-            .field("ledger_lock", &self.ledger_lock)
+            //.field("ledger_lock", &self.ledger_lock)
```
No, it does not. I'll clean that up.

tobz commented:
I went through and removed all of the "Vector only supports ..." messages that I previously used in calls to `expect(...)` as, looking at some of your comments, I came to agree that some of them were only technically correct, and not terribly insightful. These messages are either rewritten to reflect what they're doing, or what expected invariant is actually being broken... or I changed them to `unwrap` because they were in the test code and we can suss that out without some well-formatted message.

tobz commented on lib/vector-buffers/src/variants/disk_v2/common.rs:
```diff
@@ -43,6 +48,25 @@ pub(crate) fn create_crc32c_hasher() -> Hasher {
+    } else {
+        // Pad ourselves to meet the alignment requirements.
+        (amount & !(SERIALIZER_ALIGNMENT - 1)) + SERIALIZER_ALIGNMENT
+    }
```
Ah yeah, gotcha.

At any rate, I used the first approach, but ended up adding some tests for `align16`, along with better code comments and more stringent preconditions.

Addressed in f075405d5.

bruceg commented on lib/vector-buffers/src/variants/disk_v2/common.rs:
```diff
@@ -1,23 +1,28 @@
 pub const DEFAULT_MAX_RECORD_SIZE: usize = DEFAULT_MAX_DATA_FILE_SIZE;
 
+// The maximum record size has to be bigger than the record header itself, since we count the record header towards
+// sizing/space usage, etc... but we also use the overaligned version here to make sure we;re similarly accounting for
```
nit: "we;re"

bruceg commented on lib/vector-buffers/src/variants/disk_v2/common.rs:
```diff
@@ -39,10 +44,57 @@ pub const MAX_FILE_ID: u16 = u16::MAX;
+        "`amount` must be less than `MAX_ALIGNABLE_AMOUNT`"
+    );
+
+    ((amount + SERIALIZER_ALIGNMENT - 1) / SERIALIZER_ALIGNMENT) * SERIALIZER_ALIGNMENT
```
I just checked the assembly output for this, and LLVM translates this internally to `add 15` then `and ~15`. The other variant I suggested translates to the same. Smart.

tobz replied:
Yeah, I figured that the comments in the unit/property test -- avoiding depending on round to zero behavior, etc -- were going to be more of a very niche cover-our-ass situation, and that LLVM would do something fairly smart/well-defined given the power-of-two nature of the alignment amount. Good to know it's _definitely_ doing that at the moment.

bruceg commented on lib/vector-buffers/src/variants/disk_v2/common.rs:
```diff
@@ -39,10 +44,57 @@ pub const MAX_FILE_ID: u16 = u16::MAX;
 pub const MAX_FILE_ID: u16 = 6;
 
+// The alignment used by the record serializer.
+const SERIALIZER_ALIGNMENT: usize = 16;
+const MAX_ALIGNABLE_AMOUNT: usize = usize::MAX - SERIALIZER_ALIGNMENT;
```
I have a bit of a doubt about these types (and `align16` below). If these are all on-disk sizes, and we allow >4GB buffers, should they be `u64` instead of `usize`?

tobz replied:
Frankly, yeah, I'm so annoyed by having to switch between `usize` and `u64`, but...

Our usage of `rkyv`, and its overaligned vector type, `AlignedVec`, means we end up overaligning anything we serialize via `rkyv` to 16 bytes. In order to calculate the correct amount to subtract from buffer limits, we use `align16` exclusively for getting the overaligned size of both the record header itself (everything we wrap around the encoded event payload) and the ledger state, which we also use `rkyv` to handle (de)serialization for.

Since `std::mem::size_of<T>` gives us back a `usize`, like most methods that return the length of a data structure, etc, it just made sense to use `usize` with `align16`. 

Practically speaking, our size limits -- both for the buffer overall and for a given data file -- work on `u64`, so we can theoretically make them bigger than 4GB but right now we're hard-coded down to 128MB for the maximum data file size and `u64::MAX` for the buffer size, save for a user overriding that with `buffer.max_size`. Records, however, cannot be larger than `usize` just from the perspective of: we encode events to a `Vec<T>`, which can only describe its length in `usize`, and so we can't actually serialize a single record larger than `usize`.

This is all a long way of saying that we specifically try to deal with `usize` only where absolutely necessary.

bruceg reviewed (commented):
I left a comment about `usize` vs `u64` typing of values, but it maybe should be a follow-up issue.

blt commented on lib/vector-buffers/proptest-regressions/variants/disk_v2/tests/model/mod.txt:
```diff
@@ -10,3 +10,6 @@ cc 80077436fe364525dcab4e5878d617db14fae6e23a7e79f820ed592ce4614a8d # shrinks to
 cc 59a105fb23544ce39ad8e81f86628618eca77fb34143ff0c4fdd15640ad37705 # shrinks to config = DiskBufferConfig { data_dir: "/tmp/vector-disk-v2-model", max_buffer_size: 154624, max_data_file_size: 2, max_record_size: 42615, write_buffer_size: 61440, flush_interval: 10s, filesystem: TestFilesystem { inner: Mutex { data: FilesystemInner { files: {} } } } }, actions = [FlushWrites, WriteRecord(Record {  ...
 cc 4b521a9a467f44e9cbc30d830624c03605522f0310e3e972409ac7f0c42d20cd # shrinks to config = DiskBufferConfig { data_dir: "/tmp/vector-disk-v2-model", max_buffer_size: 64, max_data_file_size: 2, max_record_size: 1714, write_buffer_size: 61440, flush_interval: 10s, filesystem: TestFilesystem { inner: Mutex { data: FilesystemInner { files: {} } } } }, actions = [WriteRecord(Record { id: 0, size: 5, ev ...
 cc 99f23327ef7e759a9cb6424a3fa3495aaf4f1c7ef23145b5f5d72eb6cc5e0173 # shrinks to config = DiskBufferConfig { data_dir: "/tmp/vector-disk-v2-model", max_buffer_size: 64, max_data_file_size: 2, max_record_size: 61398, write_buffer_size: 61440, flush_interval: 10s, filesystem: TestFilesystem { inner: Mutex { data: FilesystemInner { files: {} } } } }, actions = [WriteRecord(Record { id: 0, size: 6138 ...
+cc 54be1ee096dc5169013a1cd8a114f23c5aca7a209a663d783c7f74b4c4ff4746 # shrinks to mut config = DiskBufferConfig { data_dir: "/tmp/vector-disk-v2-model", max_buffer_size: 153516, max_data_file_size: 14228, max_record_size: 14214, write_buffer_size: 61440, flush_interval: 10s, filesystem: TestFilesystem { inner: Mutex { data: FilesystemInner { files: {} } } } }, actions = [WriteRecord(Record { id: 0 ...
+cc d812fdee8da4aae904579f08cadae1b585944da58040cad4f18544a87faf240e # shrinks to mut config = DiskBufferConfig { data_dir: "/tmp/vector-disk-v2-model", max_buffer_size: 53736, max_data_file_size: 46808, max_record_size: 28912, write_buffer_size: 61440, flush_interval: 10s, filesystem: TestFilesystem { files: {} } }, actions = [WriteRecord(Record { id: 0, size: 23565, event_count: 1, aligned_len:  ...
+cc b08fb47ac81e9148d8dc6a4d3332e92c21751685be9da9bd0c5d962ad7436285 # shrinks to mut config = DiskBufferConfig { data_dir: "/tmp/vector-disk-v2-model", max_buffer_size: 69602, max_data_file_size: 2462, max_record_size: 2450, write_buffer_size: 61440, flush_interval: 10s, filesystem: TestFilesystem { files: {} } }, actions = [WriteRecord(Record { id: 0, size: 2389, event_count: 1, archived_len: 24 ...
```
This might be might favorite thing about proptest, the regression files.

tobz replied:
Big agree.

blt commented on lib/vector-buffers/src/test/common/messages.rs:
```diff
@@ -90,7 +90,7 @@ impl FixedEncodable for Message {
 
-#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
 pub(crate) struct SizedRecord(pub u32);
```
Unsure if this will work, but I think it might be a slightly tighter scope:

```suggestion
pub(crate) struct SizedRecord(pub(crate) u32);
```

blt commented on lib/vector-buffers/src/test/common/messages.rs:
```diff
@@ -116,7 +116,11 @@ impl FixedEncodable for SizedRecord {
-                "not enough capacity to encode record",
+                format!(
+                    "not enough capacity to encode record: need {}, only have {}",
+                    self.0 + 4,
```
4 for the `u32`? Might be worth pulling the `size_of` for self.0 instead.

tobz replied:
Yeah, totally fair point. I'll make this a bit more robust.

blt commented on lib/vector-buffers/src/test/common/messages.rs:
```diff
@@ -133,6 +137,10 @@ impl FixedEncodable for SizedRecord {
     }
+
+    fn encoded_size(&self) -> Option<usize> {
+        Some(self.0 as usize + 4)
```
Same comment/question here.

tobz replied:
Yeah, totally fair point. I'll make this a bit more robust.

blt commented on lib/vector-buffers/src/variants/disk_v2/tests/size_limits.rs:
```diff
@@ -377,44 +393,33 @@ async fn writer_try_write_returns_when_buffer_is_full() {
+            // Create our buffer with an arbitrarily low max buffer size, such that we can construct
+            // two writes that would otherwise fit by themselves but will end up with the second not
+            // being able to fit as the buffer is exactly full.
+            let write_size = 96;
```
Might be interesting to let proptest vary this write_size as well.

tobz replied:
This ends up happening already in the main disk v2 model, where we vary all configuration parameters (max buffer size, max data file size, and max record size, and write buffer size) as well as the size of records themselves.

blt reviewed (approved):
Solid work here. I have some small comment/questions but nothing blocking.

tobz commented:
I've made all the changes, locally, that you suggested/asked about, but I'm going to do them in a follow-up PR because I'm looking to merge this ASAP so I can try to fix the breakage it will inevitably cause in Bruce's buffer-related PR (#13396).

jszwedko commented:
> I've made all the changes, locally, that you suggested/asked about, but I'm going to do them in a follow-up PR because I'm looking to merge this ASAP so I can try to fix the breakage it will inevitably cause in Bruce's buffer-related PR (#13396).

@tobz should I plan on cherry-picking the follow-up PR for 0.23.0 too?

tobz commented:
No, the "follow-up PR" is just gonna be me committing to Bruce's existing PR. The changes will just be code cleanliness, not actual functional changes.
